### DISCUSSION TITLE
remark-math: Using \eqref and \ref tags to refer to equations BEFORE they are tagged

### DISCUSSION THREAD

**Discussion by ChauhanT** [remarkjs/remark#1211] [Q&A]
Hi,

I am trying to refer to an equation before it is tagged. So something like: 

file: test.md
```
Referring to equation $\eqref{eq-1}$ BEFORE tagging. 

$$
\begin{equation}
\label{eq-1}
\sin(x) = \sum_{n=0}^{\infty} \frac{(-1)^n}{(2n+1)!} \, x^{2n}
\end{equation}
$$

Referring to equation $\eqref{eq-1}$ AFTER tagging.
```

When rendered, I get a `(???)` from the first `\eqref{}`, and the expected `(1)` from the second one. 

The options I pass to the `rehypeMathjaxChtml` function are:

```
const MathJaxOptions = {
    chtml: {fontURL: 'https://cdn.jsdelivr.net/npm/mathjax@3/es5/output/chtml/fonts/woff-v2'},
    tex: { tags: 'ams' }
}
```

I think a double-pass might be helpful here, with the first pass generating the anchor-links and the second pass filling in the references - but this will require some sort of state control. Any idea how I can accomplish this ? If there are any scientists using `remark-mathjax`, I am sure you have already figured this one out as it seems like a common enough scenario. Any help would be greatly appreciated ! 

Thanks very much!

**Comment by wooorm** [comment]
Hey!

Math, like code, is supposed to be separate things, I think.

I also don’t maintain mathjax. I have no idea how they work regularly.

How does your idea work when you’re not using these plugins, but are just using mathjax regularly, do different containers with different math stuff share state?

I can see why not sharing state is also beneficial. What if different blocks define the same identifier?

**Reply by ChauhanT** [reply]
Hi @wooorm  ! Thanks for the response. 

I was able to do two passes in mathjax2 based on advice I got [here](https://github.com/mathjax/MathJax-node/issues/440#issuecomment-750884486). This involved sharing states between two passes, but I do not really know why/how this works. 

I have been having a discussion [here](https://github.com/mathjax/MathJax-demos-node/issues/40#issuecomment-884847111) on how to do the same with mathjax3. I think mathjax3 API has changed significantly when compared to mathjax2. I am a hobbyist, so I am not able to develop a solution for this from the ground up :( This is the reason I turned to this forum for advice and suggestions.

**Reply by wooorm** [reply]
two pases seems like a bad idea? That‘s seems slow? And if 2, why not 3? 4?

> This involved sharing states between two passes

If that works, I think I’m open to a PR to have that.

> I think mathjax3 API has changed significantly when compared to mathjax2

Hmm, well `rehype-mathjax` uses mathjax v3. If they removed support for doing that, we can’t add it back?

---

Note that sharing state will likely have tough problems. MathJax normally runs *inside one document* We run on many different documents, in arbitrary order.

So if we start using some mathjax global state, things likely break, or we need to reset things.

**Reply by ChauhanT** [reply]
The first pass sets up the equation numbers for the display math. The second pass can then fill in the `\eqref` fields. Why would you do any more passes after that ? 

I do understand you can't add it back ? 

I will try to see if I can put a PR together. I do find it very interesting that no scientist using rehype-mathjax has run into this issue before.

**Reply by wooorm** [reply]
> Why would you do any more passes after that?

Because LaTeX has an infinity “problem”/feature: everything can change the more you run things.

> I do understand you can't add it back ?

What?

> I do find it very interesting that no scientist using rehype-mathjax has run into this issue before.

I dunno. Never had this problem. Generally I’d just say: don’t depend on the order of things and managing state. Define your things in the blocks that use them.
